### ISSUE TITLE
SC2004 False positive when indexed arrays are used in arithmetic expressions

### ISSUE THREAD

**Issue by federicobond** [koalaman/shellcheck#1090]
#### For bugs
- Rule Id: SC2004 
- My shellcheck version (`shellcheck --version` or "online"): 0.4.6
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/usr/bin/env bash

array=(1 2 3)
array[2]=$((${array[2]} - 1))
```

#### Here's what shellcheck currently says:
```
In test2.sh line 4:
array[2]=$((${array[2]} - 1))
            ^-- SC2004: $/${} is unnecessary on arithmetic variables.
```
#### Here's what I wanted or expected to see:

No error, since the braces are necessary to prevent Bash from erroring out with:
```
line 4: 1[2] - 1: syntax error: invalid arithmetic operator (error token is "[2] - 1")
```

**Comment by dguyton** [comment]
Highlighting a slightly different use case where there ought to be a relevant SC2004 warning, but shellcheck reports the possible issue in the wrong part of the line. It's an extenuation of @federicobond's use case, with a twist.

Here's an example. I'm using shellcheck 0.6.0 with BASH 4.3.48(1).

```
In my-program.sh line 1195:
        if [ "$((${hightrack[cdisc]}))" -gt 0 ]; then # 1/ skip if current disc has no recorded track numbers (hightrack for current disc = 0)
                 ^-----------------^ SC2004: $/${} is unnecessary on arithmetic variables.
```
What is interesting is if I make the bracketed [ ] name a string, the SC2004 comment goes away and I don't get a false positive - even though the other portion of the line is very similar to the original use case.

In this particular case I actually had an error in the line, which I noticed because of shellcheck. Not a syntax error, but a logic error on my part. Specifically, the last var reference (cdisc) should have been ($cdisc). So, my point is there's value here in this scenario for SC2004 if were tweaked to prompt the user to verify their code (my use case), but the use case reported by @federicobond is still true. The SC2004 comment I receive is initiated by his use case, not mine. The former is not relevant and shellcheck should not report it. The latter (my) variation is relevant, but SC2004 does not correctly pinpoint the potential issue.

**Comment by jabberabbe** [comment]
And @federicobond, what about `array[2]=$((array[2] - 1))`? I think that ShellCheck is suggesting to remove the braces _and_ the dollar sign.

**Comment by koalaman** [comment]
@federicobond You do indeed have to remove the whole `${` and `}`, not just the `$`. I agree that the message is not very explicit about this.

It really just says `$/${}` instead of explicitly `${}` for the lazy technical reason that at this point in the processing, the information about whether or not `var` was referenced as `$var` or `${var}` is no longer available. 

The [wiki page for SC2004](https://github.com/koalaman/shellcheck/wiki/SC2004) includes this as an example, but your version v0.4.6 offers no help in finding this additional information. v0.6.0+ will show relevant wiki links at the end of the output.

@dguyton This is actually a weird edge case. `${..}` *is* required for associative arrays in some instances, such as with `index="]"`. I filed #1481 about making this inconsistency less strange.
